src/resolve/resolved-bus.c

```diff
@@ -360,10 +362,53 @@ static int parse_as_address(sd_bus_message *m, int ifindex, const char *hostname
         return sd_bus_send(sd_bus_message_get_bus(m), reply, NULL);
 }
 
+static int bus_check_dns_access(Manager *m, sd_bus_message *message, const char *hostname, char **ret_unit) {
+        int r = 0;
+        _cleanup_(sd_bus_creds_unrefp) sd_bus_creds *creds = NULL, *pid_creds = NULL;
+        const char *unit = NULL;
+        pid_t pid;
+
+        assert(hostname);
+
+        /* Get creds from message */
+        r = sd_bus_query_sender_creds(message, SD_BUS_CREDS_PID, &creds);
+        if (r < 0) {
+                log_debug("Couldn't query sender creds");
+                return 0;
+        }
+
+        r = sd_bus_creds_get_pid(creds, &pid);
+        if (r < 0) {
+                log_debug("Couldn't get PID from creds");
+                return 0;
+        }
+
+        /* Get other creds from sender's PID */
+        r = sd_bus_creds_new_from_pid(&pid_creds, pid, SD_BUS_CREDS_UNIT);
+        if (r < 0) {
+                log_debug("Couldn't query creds from sender's PID");
+                return 0;
+        }
+
+        r = sd_bus_creds_get_unit(pid_creds, &unit);
+        if (r < 0 || !unit) {
+                log_debug("Couldn't get unit for pid %d", pid);
+                return 0;
+        }
+
+        r = manager_check_dns_access(m, hostname, unit);
+        if (r < 0)
+                return r;
+
+        *ret_unit = strdup(unit);
+        return 0;
+}
+
 static int bus_method_resolve_hostname(sd_bus_message *message, void *userdata, sd_bus_error *error) {
         _cleanup_(dns_question_unrefp) DnsQuestion *question_idna = NULL, *question_utf8 = NULL;
         Manager *m = userdata;
         const char *hostname;
+        char *unit = NULL;
```

Comment:
_cleanup_free_ here and TAKE_PTR assigning to q->unit ? Otherwise strdup'ed string ref get leaked if there's a failure between bus_check_dns_access and the assigment to q

Reply:
Right, fixed.